app/controllers/triannon/auth_controller.rb

```diff
@@ -0,0 +1,441 @@
+
+    # Issue a 403 for invalid access token
+    def access_token_invalid
+      err = {
+        error: 'invalidAccess',
+        errorDescription: 'invalid access token',
+        errorUri: ''
+      }
+      json_response(err, 403)
+    end
+
+    # Issue a 401 to challenge for a client access token
+    def access_token_required
+      err = {
+        error: 'invalidAccess',
+        errorDescription: 'access token is required',
+        errorUri: ''
+      }
+      json_response(err, 401)
+    end
+
+
+    # --------------------------------------------------------------------
+    # User authentication
+
+    # Handles POST /auth/login
+    # The request MUST include a URI parameter 'code=client_token' where
+    # the 'client_token' has been obtained from /auth/client_identity and
+    # the request MUST carry a body with the following JSON template:
+    # { "userId" : "ID", "userSecret" : "SECRET", "workgroups" : "wgA, wgB" }
+    def login_handler_post
+      return unless process_post?
+      return unless process_json?
+      auth_code = params[:code]
+      if auth_code.nil?
+        auth_code_required
+      elsif auth_code_valid?(auth_code)
+        data = JSON.parse(request.body.read)
+        required_fields = ['userId', 'userSecret', 'workgroups']
+        identity = parse_identity(data, required_fields)
+        # When an authorized client POSTs user data, it is simply accepted.
+        if identity['userId'] && identity['workgroups']
+          cookies[:login_user] = identity['userId']
+          session[:login_data] = identity
+          redirect_to root_url, notice: 'Successfully logged in.'
+        else
+          login_required
+        end
+      else
+        auth_code_invalid
+      end
+    end
+
+    def login_required
+      if request.format == :html
+        request_http_basic_authentication
+      elsif request.format == :json
+        response.headers["WWW-Authenticate"] = %(Basic realm="Application")
+        err = {
+          error: '401 Unauthorized',
+          errorDescription: 'login credentials required',
+          errorUri: 'http://image-auth.iiif.io/api/image/2.1/authentication.html'
+        }
+        json_response(err, 401)
+      end
+    end
+
+
+    # --------------------------------------------------------------------
+    # Client authentication
+
+    # Authenticates known clients
+    # @param identity [Hash] with fields 'clientId' and 'clientSecret'
+    def authorized_client?(identity)
+      authorized_clients[identity['clientId']] == identity['clientSecret']
+    end
+
+    # Sets a hash of authorized client data key:value pairs that correspond to
+    # the client_identity service parameters named 'clientId':'clientSecret';
+    # the data is provided by configuration.
+    def authorized_clients
+      @authorized_clients ||= Triannon.config[:authorized_clients]
+    end
+
+
+    # --------------------------------------------------------------------
+    # Authentication tokens
```

Comment:
The app controller should make available the #auth_token_data method to all controllers.  The auth controller handles all the auth requests.  These methods in the auth controller are for an authorized client to authenticate.  They are not related to the app controller methods, which are more closely related to the 'user' authentication (which is done by an authenticated client).